meta-ros2/classes/ros_ament_cmake.bbclass

```diff
@@ -25,11 +25,13 @@ export AMENT_PREFIX_PATH="${STAGING_DIR_HOST}${prefix};${STAGING_DIR_NATIVE}${pr
 
 inherit cmake python3native
 
+PYTHONPATH:prepend:class-nativesdk = "${STAGING_DIR_NATIVE}/opt/ros/${ROS_DISTRO}/lib/python3.12/site-packages:"
```

Comment:
1. Use ${ros_prefix} to replace opt/ros/${ROS_DISTRO}
2. Use ${PYTHON_BASEVERSION} instead of python3.12 is better.

Reply:
Thanks! I will test and feedback. Sorry for the delay, my attention has been distracted.

Reply:
Using 1. fails to build the SDK with the following error for the packages `rosidl-typesupport-interface`, `rosidl-cmake`, `rcutils` and `rosidl-generator-type-description`

```
| DEBUG: Python function extend_recipe_sysroot finished
| DEBUG: Executing shell function do_configure
| -- Found ament_cmake: 2.6.0 (/home/stephen/Workspace/yocto/pi5/build/tmp-glibc/work/x86_64-nativesdk-oesdk-linux/nativesdk-rosidl-cmake/4.7.0-1/recipe-sysroot-native/opt/ros/rolling/share/ament_cmake/cmake)
| -- Found Python3: /home/stephen/Workspace/yocto/pi5/build/tmp-glibc/work/x86_64-nativesdk-oesdk-linux/nativesdk-rosidl-cmake/4.7.0-1/recipe-sysroot-native/usr/bin/python3-native/python3 (found version "3.12.3") found components: Interpreter
| Traceback (most recent call last):
|   File "/home/stephen/Workspace/yocto/pi5/build/tmp-glibc/work/x86_64-nativesdk-oesdk-linux/nativesdk-rosidl-cmake/4.7.0-1/recipe-sysroot-native/opt/ros/rolling/share/ament_cmake_core/cmake/package_templates/templates_2_cmake.py", line 21, in <module>
|     from ament_package.templates import get_environment_hook_template_path
| ModuleNotFoundError: No module named 'ament_package'
| CMake Error at /home/stephen/Workspace/yocto/pi5/build/tmp-glibc/work/x86_64-nativesdk-oesdk-linux/nativesdk-rosidl-cmake/4.7.0-1/recipe-sysroot-native/opt/ros/rolling/share/ament_cmake_core/cmake/ament_cmake_package_templates-extras.cmake:41 (message):
| 
|   execute_process(/home/stephen/Workspace/yocto/pi5/build/tmp-glibc/work/x86_64-nativesdk-oesdk-linux/nativesdk-rosidl-cmake/4.7.0-1/recipe-sysroot-native/usr/bin/python3-native/python3
|   /home/stephen/Workspace/yocto/pi5/build/tmp-glibc/work/x86_64-nativesdk-oesdk-linux/nativesdk-rosidl-cmake/4.7.0-1/recipe-sysroot-native/opt/ros/rolling/share/ament_cmake_core/cmake/package_templates/templates_2_cmake.py
|   /home/stephen/Workspace/yocto/pi5/build/tmp-glibc/work/x86_64-nativesdk-oesdk-linux/nativesdk-rosidl-cmake/4.7.0-1/build/ament_cmake_package_templates/templates.cmake)
|   returned error code 1
| Call Stack (most recent call first):
|   /home/stephen/Workspace/yocto/pi5/build/tmp-glibc/work/x86_64-nativesdk-oesdk-linux/nativesdk-rosidl-cmake/4.7.0-1/recipe-sysroot-native/opt/ros/rolling/share/ament_cmake_core/cmake/ament_cmake_coreConfig.cmake:41 (include)
|   /home/stephen/Workspace/yocto/pi5/build/tmp-glibc/work/x86_64-nativesdk-oesdk-linux/nativesdk-rosidl-cmake/4.7.0-1/recipe-sysroot-native/opt/ros/rolling/share/ament_cmake/cmake/ament_cmake_export_dependencies-extras.cmake:15 (find_package)
|   /home/stephen/Workspace/yocto/pi5/build/tmp-glibc/work/x86_64-nativesdk-oesdk-linux/nativesdk-rosidl-cmake/4.7.0-1/recipe-sysroot-native/opt/ros/rolling/share/ament_cmake/cmake/ament_cmakeConfig.cmake:41 (include)
|   CMakeLists.txt:5 (find_package)
```

Reply:
As a corollary, remove `${STAGING_DIR_NATIVE}/opt/ros/${ROS_DISTRO}` from `EXTRA_OECMAKE:prepend:class-nativesdk -D-DCMAKE_PREFIX_PATH` line results in similar errors:
```
| CMake Error at CMakeLists.txt:5 (find_package):
|   By not providing "Findament_cmake.cmake" in CMAKE_MODULE_PATH this project
|   has asked CMake to find a package configuration file provided by
|   "ament_cmake", but CMake did not find one.
| 
|   Could not find a package configuration file provided by "ament_cmake" with
|   any of the following names:
| 
|     ament_cmakeConfig.cmake
|     ament_cmake-config.cmake
| 
|   Add the installation prefix of "ament_cmake" to CMAKE_PREFIX_PATH or set
|   "ament_cmake_DIR" to a directory containing one of the above files.  If
|   "ament_cmake" provides a separate development package or SDK, be sure it
|   has been installed.
```
for `rosidl/rosidl-typesupport-interface` and `rosidl-cmake` while the following error generated for `rcutils`:
```
| CMake Error at CMakeLists.txt:16 (find_package):
|   By not providing "Findament_cmake_python.cmake" in CMAKE_MODULE_PATH this
|   project has asked CMake to find a package configuration file provided by
|   "ament_cmake_python", but CMake did not find one.
| 
|   Could not find a package configuration file provided by
|   "ament_cmake_python" with any of the following names:
| 
|     ament_cmake_pythonConfig.cmake
|     ament_cmake_python-config.cmake
| 
|   Add the installation prefix of "ament_cmake_python" to CMAKE_PREFIX_PATH or
|   set "ament_cmake_python_DIR" to a directory containing one of the above
|   files.  If "ament_cmake_python" provides a separate development package or
|   SDK, be sure it has been installed.
```

Reply:
The appear to be a problem with the ${ros_prefix} value.

Reply:
Using 2. generates the variation of the following error for packages `rosidl-typesupport-interface`, `rosidl-cmake`, `rcutils`, `rosidl-parser` and `rosidl-generator-type-description`:
```
| DEBUG: Python function extend_recipe_sysroot finished
| DEBUG: Executing shell function do_configure
| -- Found ament_cmake: 2.6.0 (/home/stephen/Workspace/yocto/pi5/build/tmp-glibc/work/x86_64-nativesdk-oesdk-linux/nativesdk-rosidl-typesupport-interface/4.7.0-1/recipe-sysroot-native/opt/ros/rolling/share/ament_cmake/cmake)
| -- Found Python3: /home/stephen/Workspace/yocto/pi5/build/tmp-glibc/work/x86_64-nativesdk-oesdk-linux/nativesdk-rosidl-typesupport-interface/4.7.0-1/recipe-sysroot-native/usr/bin/python3-native/python3 (found version "3.12.3") found components: Interpreter
| Traceback (most recent call last):
|   File "/home/stephen/Workspace/yocto/pi5/build/tmp-glibc/work/x86_64-nativesdk-oesdk-linux/nativesdk-rosidl-typesupport-interface/4.7.0-1/recipe-sysroot-native/opt/ros/rolling/share/ament_cmake_core/cmake/package_templates/templates_2_cmake.py", line 21, in <module>
|     from ament_package.templates import get_environment_hook_template_path
| ModuleNotFoundError: No module named 'ament_package'
| CMake Error at /home/stephen/Workspace/yocto/pi5/build/tmp-glibc/work/x86_64-nativesdk-oesdk-linux/nativesdk-rosidl-typesupport-interface/4.7.0-1/recipe-sysroot-native/opt/ros/rolling/share/ament_cmake_core/cmake/ament_cmake_package_templates-extras.cmake:41 (message):
| 
|   execute_process(/home/stephen/Workspace/yocto/pi5/build/tmp-glibc/work/x86_64-nativesdk-oesdk-linux/nativesdk-rosidl-typesupport-interface/4.7.0-1/recipe-sysroot-native/usr/bin/python3-native/python3
|   /home/stephen/Workspace/yocto/pi5/build/tmp-glibc/work/x86_64-nativesdk-oesdk-linux/nativesdk-rosidl-typesupport-interface/4.7.0-1/recipe-sysroot-native/opt/ros/rolling/share/ament_cmake_core/cmake/package_templates/templates_2_cmake.py
|   /home/stephen/Workspace/yocto/pi5/build/tmp-glibc/work/x86_64-nativesdk-oesdk-linux/nativesdk-rosidl-typesupport-interface/4.7.0-1/build/ament_cmake_package_templates/templates.cmake)
|   returned error code 1
| Call Stack (most recent call first):
|   /home/stephen/Workspace/yocto/pi5/build/tmp-glibc/work/x86_64-nativesdk-oesdk-linux/nativesdk-rosidl-typesupport-interface/4.7.0-1/recipe-sysroot-native/opt/ros/rolling/share/ament_cmake_core/cmake/ament_cmake_coreConfig.cmake:41 (include)
|   /home/stephen/Workspace/yocto/pi5/build/tmp-glibc/work/x86_64-nativesdk-oesdk-linux/nativesdk-rosidl-typesupport-interface/4.7.0-1/recipe-sysroot-native/opt/ros/rolling/share/ament_cmake/cmake/ament_cmake_export_dependencies-extras.cmake:15 (find_package)
|   /home/stephen/Workspace/yocto/pi5/build/tmp-glibc/work/x86_64-nativesdk-oesdk-linux/nativesdk-rosidl-typesupport-interface/4.7.0-1/recipe-sysroot-native/opt/ros/rolling/share/ament_cmake/cmake/ament_cmakeConfig.cmake:41 (include)
|   CMakeLists.txt:5 (find_package)
```

Reply:
Investigating value of '${ros_prefix}' for the native-sdk
```
$ bitbake-getvar -r nativesdk-rosidl-typesupport-interface --value ros_prefix
/usr/local/oe-sdk-hardcoded-buildpath/sysroots/x86_64-oesdk-linux/opt/ros/rolling
```
while for the target:
```
$ bitbake-getvar -r rosidl-typesupport-interface --value ros_prefix
/opt/ros/rolling
```

Reply:
Search meta-ros for `ros_prefix` yields the following definition in `meta-ros-common/classes/ros_opt_prefix.bbclass`:
```
ros_prefix ?= "${base_prefix}/opt/ros/${ROS_DISTRO}"
```
I did not find (could be wrong here) any other prior definitions of `ros_prefix` so `base_prefix` must be the governing variable. I found `base_prefix` in `openembedded-core/meta/classes-recipe/nativesdk.bbclass`:
```
base_prefix = "${SDKPATHNATIVE}"
```
Tracking `SDKPATHNATIVE` to `openembedded-core/meta/conf/bitbake.conf` the definition is
```
SDK_SYS = "${SDK_ARCH}${SDK_VENDOR}-${SDK_OS}"
SDKPATH = "/usr/local/oe-sdk-hardcoded-buildpath"
SDKPATHNATIVE = "${SDKPATH}/sysroots/${SDK_SYS}"
```
Which starts to make sense in explaining the observed behavior.  I think that the current value of `ros_prefix` is wrong for building the SDK. Unfortunately I'm not sure the correct way to fix this would be?

Reply:
Thanks for your details.  I didn't realize the prefix_xx are changed in nativesdk.bbclass.
```
base_prefix = "${SDKPATHNATIVE}"
prefix = "${SDKPATHNATIVE}${prefix_nativesdk}"
```
My intention is to use the variables defined in ros_opt_prefix.bbclass.

Reply:
That seems right, may be we need to add an variable with it's default value as `/opt/ros/${ROS_DISTRO}`. This would allow the some flexibility when construction paths for the nativesdk.